### ISSUE TITLE
Supported formats could not be retrieved because an error occurred connecting to the camera service

### ISSUE THREAD

**Issue by rishabhja** [twilio/video-quickstart-android#284]
### Description

App crashes due to the following error log. It did not crash before beta 5 version. 

### Steps to Reproduce

1. Give permissions to camera.
2. App crashes when local camera stream is created.

#### Logs

```
java.lang.RuntimeException: Unable to resume activity {to.go_internal/to.go.app.twilio.VideoConferenceActivity}: java.lang.IllegalStateException: Supported formats could not be retrieved because an error occurred connecting to the camera service
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3486) ~[na:0.0]
	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3526) ~[na:0.0]
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1584) ~[na:0.0]
	at android.os.Handler.dispatchMessage(Handler.java:102) ~[na:0.0]
	at android.os.Looper.loop(Looper.java:163) ~[na:0.0]
	at android.app.ActivityThread.main(ActivityThread.java:6321) ~[na:0.0]
	at java.lang.reflect.Method.invoke(Native Method) ~[na:0.0]
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:880) ~[na:0.0]
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:770) ~[na:0.0]
Caused by: java.lang.IllegalStateException: Supported formats could not be retrieved because an error occurred connecting to the camera service
	at com.twilio.video.Preconditions.checkState(Preconditions.java:453) ~[na:0.0]
	at com.twilio.video.CameraCapturer.getSupportedFormats(CameraCapturer.java:347) ~[na:0.0]
	at com.twilio.video.LocalVideoTrack.create(LocalVideoTrack.java:154) ~[na:0.0]

```

#### Expected Behavior

App should not crash

#### Actual Behavior

App crashes

#### Reproduces how Often

100%

#### Video Android SDK

2.0.0 beta 5

#### Android API

23-27

#### Android Device

One Plus, Xiaomi

**Comment by aaalaniz** [comment]
Hey @rishabhja 

There were no changes to the camera capturers in this release, so this is a little surprising. Can you provide some steps to reproduce? I typically see this error when trying to create a video track with a camera capturer while the camera is in use.

Thanks!

**Comment by rishabhja** [comment]
Yes. So basically, when user tried to join the room in a freshly installed app, this crash occurs as soon as he joins the room after giving the camera permission. It does not occur after that

**Comment by aaalaniz** [comment]
The log states that you are creating a video track in onResume. I recommend creating the local video track with camera capturer in `onRequestPermissionsResult` as seen [in our quickstart](https://github.com/twilio/video-quickstart-android/blob/master/quickstart/src/main/java/com/twilio/video/quickstart/activity/VideoActivity.java) and checking that permission is available in onResume before [creating the track](https://github.com/twilio/video-quickstart-android/blob/master/quickstart/src/main/java/com/twilio/video/quickstart/activity/VideoActivity.java#L270).

**Comment by rishabhja** [comment]
I am checking for permissions in onResume before creating the track and I am requesting the permission in onCreate. I do not start the video track until the permission has been granted

**Comment by aaalaniz** [comment]
Is it possible a video track is being created twice? Once in onRequestPermissionsResult and once in onResume or are you only creating the track in one of those callbacks?

**Comment by rishabhja** [comment]
I think that is possible. Can that be a cause of the crash?

**Comment by aaalaniz** [comment]
Yes. I recommend creating only one video track with a camera capturer. When the track is created the camera resource is acquired. If another track with camera capturer is created then this error would occur because the camera is in use by the other track.

**Comment by rishabhja** [comment]
This happens even after the fix you prooosed.

**Comment by aaalaniz** [comment]
Hey @rishabhja 

Is it possible to reproduce this bug in the quickstart?

Thanks!

**Comment by rishabhja** [comment]
Also facing this on LG G2:

`FATAL EXCEPTION: Camera SurfaceTextureHelper
                                                                Process: to.go_internal, PID: 19374
                                                                java.lang.RuntimeException: getParameters failed (empty parameters)
                                                                    at android.hardware.Camera.native_getParameters(Native Method)
                                                                    at android.hardware.Camera.getParameters(Camera.java:1965)
                                                                    at org.webrtc.Camera1Session.create(Camera1Session.java:83)
                                                                    at org.webrtc.Camera1Capturer.createCameraSession(Camera1Capturer.java:34)
                                                                    at org.webrtc.CameraCapturer$5.run(CameraCapturer.java:285)
                                                                    at android.os.Handler.handleCallback(Handler.java:739)
                                                                    at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                    at android.os.Looper.loop(Looper.java:148)
                                                                    at android.os.HandlerThread.run(HandlerThread.java:61)
`

**Comment by adityaflexsin** [comment]
this issue after first call disconnect after then call video Activity then crash on this code first line:
localVideoTrack = LocalVideoTrack.create(this,
true,
cameraCapturerCompat.getVideoCapturer(),
LOCAL_VIDEO_TRACK_NAME);
primaryVideoView.setMirror(true);
localVideoTrack.addRenderer(primaryVideoView);
localVideoView = primaryVideoView;
And this issue on every devices.

I am using audio and video track release and null onDestroy()

@Override
protected void onDestroy() {
/*
* Always disconnect from the room before leaving the Activity to
* ensure any memory allocated to the Room resource is freed.
*/
if (room != null && room.getState() != RoomState.DISCONNECTED) {
room.disconnect();
disconnectedFromOnDestroy = true;
}

    /*
     * Release the local audio and video tracks ensuring any memory allocated to audio
     * or video is freed.
     */
    if (localAudioTrack != null) {
        localAudioTrack.release();
        localAudioTrack = null;
    }
    if (localVideoTrack != null) {
        localVideoTrack.release();
        localVideoTrack = null;
    }

    super.onDestroy();
}

**Comment by aaalaniz** [comment]
hey @adityaflexsin 

Please reproduce this issue in a fork of our quickstart. We will be happy to investigate with proper reproduction steps.

Thanks!

**Comment by ducna255** [comment]
I'm facing with this issue too. Have any solution?

**Comment by april21dev** [comment]
I also facing this issue too.
first call is okay, but on second try, It crashes.
I release local video track when activity's onPause() or onDestroy() called.
I don't know why it crashes nevertheless I released the camera.
Is there delay when releasing camera?

**Comment by aaalaniz** [comment]
Hey @april21dev 

When you experience the issue do you see that the Camera app can be opened or does it crash as @ducna255 described?

**Comment by zackm0571** [comment]
Closing for now. Please reopen with steps to reproduce and a link to a fork if you continue to have this issue :)

**Comment by poojashah23996** [comment]
I am facing same issue for lollipop versio also i am getting  java.lang.IllegalStateException: Supported formats could not be retrieved because an error occurred connecting to the camera service exception for the version less than marhsmallow

**Comment by leafygreens88** [comment]
restarting my emulator fixed it for me.

**Comment by timusus** [comment]
I have this issue after exiting and then relaunching the app - it seems that the camera is not correctly released for some reason (despite calling `videoTrack.release()`. Calling `cameraCapturer.stopCapture()` when the activity is destroyed (app is exited) solved the problem for me.

I suspect there's some sort of race condition preventing `stopCapture()` from executing correctly, perhaps if you quit the app too soon after starting iniitialisation of the camera capturer, then the capturer is not released.

**Comment by aaalaniz** [comment]
Hey @timusus 

Can you file a separate issue for this with the steps your app used to reproduce? I think we can write a test for this in our SDK to reproduce and fix.

Thanks!

**Comment by mohammedhemaid** [comment]
@timusus  
I am facing the same issue as you but I am using CameraCapturerCompat and it doesn't have the stopCapture() method

**Comment by mohammedhemaid** [comment]
ok ok I found this one cameraCapturer?.videoCapturer?.stopCapture()
